mathics/core/symbols.py

```diff
@@ -807,26 +816,40 @@ def __getnewargs__(self):
         return (self.name, self.sympy_dummy)
 
 
+class Constant(Symbol, ExpandOnce):
+    """For Symbols that are cannot be changed"""
+
+    def __init__(self, name, sympy_dummy=None):
+        self = super().__new__(self, name, sympy_dummy)
+
+    pass
+
+
 # Symbols used in this module.
 
-SymbolFalse = Symbol("System`False")
-SymbolGraphics = Symbol("System`Graphics")
-SymbolGraphics3D = Symbol("System`Graphics3D")
-SymbolHoldForm = Symbol("System`HoldForm")
-SymbolList = Symbol("System`List")
-SymbolMachinePrecision = Symbol("MachinePrecision")
-SymbolMakeBoxes = Symbol("System`MakeBoxes")
-SymbolMaxPrecision = Symbol("$MaxPrecision")
-SymbolMinPrecision = Symbol("$MinPrecision")
-SymbolN = Symbol("System`N")
+SymbolFalse = Constant("System`False")
+SymbolGraphics = Constant("System`Graphics")
+SymbolGraphics3D = Constant("System`Graphics3D")
+SymbolHoldForm = Constant("System`HoldForm")
+SymbolList = Constant("System`List")
+SymbolMachinePrecision = Constant("MachinePrecision")
+SymbolMakeBoxes = Constant("System`MakeBoxes")
+SymbolMaxPrecision = Constant("$MaxPrecision")
+SymbolMinPrecision = Constant("$MinPrecision")
+SymbolN = Constant("System`N")
+
+# FIXME: LinearModelFit[{{2, 1}, {3, 4}, {5, 3}, {7, 6}}, x, x] ;
+# fails if we use a Constant below. Fix whatever is making this fail.
+# SymbolNull = Constant("System`Null")
 SymbolNull = Symbol("System`Null")
```

Comment:
Probably the most prevalent place where we do go through an unneeded recheck of a value looking for a fixed point is when the value is Null, because that is probably the most common value returned. 

I don't understand what causes a failure here, but if this were fixed it would probably have a small impact on the performance.

Reply:
It would be nice to see how the effect these changes have on the performance. Print tests, seem to indicate on startup we are making additional trips quite a bit.

Reply:
There are some `is SymbolNull` in the code, and `Constant("Symbol`Null")` is different than `Symbol("Symbol`Null")` because expressions are first evaluated to symbols, not to a specific class. A straight `is SymbolNull` doesn't convert the symbol `Null` to an instance of `Contant`. I don't believe this is a bug, but a defined behavior.

Reply:
Ok. I think I understand and this make sense.  I noticed in parsing Symbol("System`Null") was used as it should be. And it makes sense that initially values are their own distinct placeholders of Null values. 

Then we just need to make note of this information in a comment here.

Reply:
My idea about "singletonize" symbols was to make faster the "sameQ" comparisons: to check that a leaf in an expression is `System`Null` requires just to check its memory position. In principle, with this new implementation that also works (just changes the class). However, this would require an extra step of parsing/evaluating `Symbol(Null)->Constant(Null)`.

Reply:
> My idea about "singletonize" symbols was to make faster the "sameQ" comparisons: to check that a leaf in an expression is `System`Null`requires just to check its memory position. In principle, with this new implementation that also works (just changes the class). However, this would require an extra step of parsing/evaluating`Symbol(Null)->Constant(Null)`.

This idea is really part of a larger principle found in Lisp and other programing languages (Python has "intern" for example),  that strings and symbols are different.  A string can be thought of as composition of characters that can be selected and mutated. For example I can form a new string by adding letters or extracting letters. 

A symbol has the property name, and you generally don't want to work on the name other than check that it is the same as some other name. That is why in Lisp and other languages that SameQ which is modelled on Lisp's Eq works that way and is effient. It is more about the concept that affords an efficient implementation rather than the other way around. 

It is unfortunate and possibly confusing that Symbol has two meanings the Lisp meaning described above and the meaning in WL where it is a symbolic placeholder for a value. 

We should be making this distinction clearer in docstrings and comments, especially as it has caused confusion among us. 

Also what is needed is a reevaluation of what properties are needed for Symbols of the Mathematica kind and Symbols of the Lisp kind (just a placeholder for a name of a thing). We don't need all all the methods in the Lisp Symbol that Mathics Symbol currently has. 

And then there is concept of a Literal value (which also is being overloaded by the word Constant). That too has certain operations that make sense on it and some that don't make sense on it. And the ones that don't make sense but are a part of Symbol should be separated. 

This is a mess.

Reply:
@rocky, is this the notion of a Lisp symbol that you are talking about? https://www.cs.cmu.edu/Groups/AI/html/cltl/clm/node107.html#SECTION001400000000000000000